### DISCUSSION TITLE
Closest METAR Station

### DISCUSSION THREAD

**Discussion by UnsterblichMedic** [weather-gov/api#451] [Q&A]
Trying to figure out a way to get the closest station to a latitude and longitude that reports METAR.  I have a system setup on linux that uses the Text Data Server over at aviationweather.gov but it requires a wget and CORS policy prevents an API request.  So, I am trying to see if this API will accomplish the same thing.

I can get all the stations from /points/{points} that is in a list of nearest to farthest (I am pretty sure in that order).  I can also get a list of METAR stations with /products/types/MTR, but I believe that is all of them.  I would really like to just have something like /points/{points}/MTR/stations or something.  I think there is a gridpoints one but I am getting returns that are very far from me.  I am also having a hard time using the information in the @graph section of some of the API responses (new to json-ld and having hard time finding information on extracting data of these).  Any help, directions or links are welcomed.  Thank you.

**Comment by JJKraw** [comment]
It this what you mean by "I can get all the stations from /point/{point}"?

I locate the nearest METAR location by first doing /point/{point} and then fetch the url returned in the observationStations (form of /gridpoints/{wfo}/{x},{y}/stations). The first station returned is my nearest METAR source.

Does that not work for you?

**Reply by UnsterblichMedic** [reply]
That seemed to work.  I used the `/points/{point}`, then made another request using `response.data.properties.observationStations` which was `/gridpoints/STO/50,70/stations` then used `response.data.observationStations[0]` adding `/observations/latest` to make a request of `/stations/KMHR/observations/latest`.  This gave me the most recent METAR report and I accessed it with `response.data.properties.rawMessage`.  I will continue to test it in different locations to make sure it is working.  I guess I am confused on how it knows I want the closest station with a METAR report since not all of them give the METAR.  I am also a little unclear what the `{wfo}` stands for and how to figure out gridpoints.  I know how to retrieve them from the API requests but it would be nice to have a map that shows where the supplied gridpoints correlate to as I am unfamiliar with gridpoints.  Thank you so much as I am new to this API, progress!

**Reply by UnsterblichMedic** [reply]
So just checked some other locations by manually entering the coordinates.  It seemed like it was working great until I put in a rural location.  The station it pulled up did not have METAR reporting.  I need a way to filter out only stations with METAR reporting.

**Reply by JJKraw** [reply]
WFO is Weather Forecast Office.  Here is one map that shows them all.

  https://www.weather.gov/srh/nwsoffices

The gridpoints are used to identify the individual areas used for point forecasts. So if you fetch

  https://api.weather.gov/gridpoints/STO/50,70/forecast

and accept geoJson, you'll see the bounding box for the point forecast under "geometry".

Can you give the example for the rural area you tried?

**Reply by UnsterblichMedic** [reply]
Excellent, thank you. Here is an example of what I am running into.
`https://api.weather.gov/points/37.83,-120.54`
`https://api.weather.gov/gridpoints/STO/67,28/stations`
The first station is this one and it does not have a METAR. 
`https://api.weather.gov/stations/GNSC1/observations/latest`
The second station does have a METAR. The only way I can think of is to loop through the results and grab the first station that has a rawMessage for the METAR but I was hoping for a way of just making the API call. Here is the second station that has a METAR. 
`https://api.weather.gov/stations/KO22/observations/latest `

**Reply by JJKraw** [reply]
The NWS folks can correct me if I'm wrong, but I think you want to stick to the four letter stations with a first letter of K (contiguous 48 states), P (Alaska, Hawaii, other Pacific islands), T (Caribbean), M (Guantanamo). Might be others.

I'd grab the first one in the stations list of that type. These are mostly (or all) airports I believe. E.g. KBOS is Boston, KDTW is Detroit...

**Reply by UnsterblichMedic** [reply]
Yea, airports are delineated with a K in the US. I was hoping there was an API request to filter out everything else like with products or something. I will do the work around of iterating through the array for the first one that matches that. Thank you so much for your help. It is greatly appreciated.

**Comment by scadergit** [comment]
This request is too application-specific to create a cache permutation that would reduce the performance potential of the API. It would be ideal to be on the cloud with overhead to offer this level of custom queries (e.g. GraphQL), but in our current environment every request counts, where once a month we're asked to reconsider the number of permutations on the alerts endpoints.
